improve: refactor storages in PlayerStorage component and optimized save

dudantas opened the pull request:
The storage system was previously managed in Player through a raw storageMap, with direct insertions, deletions, and range generation (genReservedStorageRange). This change caused unnecessary complete rewrites on save and spread logic across Player, IOLoginData, and related components.

Changes
- Introduced a new PlayerStorage component to encapsulate all storage-related logic
- Removed the legacy storageMap from Player, centralizing load/save and manipulation in the new component
- Implemented incremental persistence: only modified or removed keys are saved to the database, avoiding complete rewrites
- Added support for reserved key ranges (outfits, familiars) within PlayerStorage::getReservedRange
- Updated IOLoginDataLoad and IOLoginDataSave to delegate load/save to PlayerStorage
- Added missing benchmark.cpp implementation and updated build files accordingly
- Removed storage name in favor of KV.

This change refactor improves code organization, maintainability, and performance. By persisting only modified keys instead of the full storage set, save operations are faster and reduce unnecessary database load.

dudantas committed:
Refactor player storage to PlayerStorage component

Moved player storage logic from Player to a new PlayerStorage component for better encapsulation and maintainability. Updated all relevant code to use the new component, including storage loading, saving, and access methods. Deprecated direct access to storageMap in Player and replaced macro IS_IN_KEYRANGE with constexpr isStorageKeyInRange function.

Added a comprehensive README.md for PlayerStorage, detailing its design, usage, reserved ranges, and performance optimizations. Enhanced player_storage.hpp with detailed Doxygen-style comments for all public methods and members, improving code readability and maintainability.

dudantas committed:
Add unit tests for PlayerStorage component

Introduces a new test suite for the PlayerStorage component under tests/unit/players/components. The tests cover adding, retrieving, and removing storage values, as well as handling reserved keys. Updates CMakeLists.txt files to include the new test directories and sources.

Add const accessors for outfits and familiars to Player

Introduced const getter methods getOutfits() and getFamiliars() in Player to provide read-only access to the respective collections. Updated unit tests to use these new accessors, improving encapsulation and ensuring side-effects are verified via the public API.

Fix missing std:: qualifier for log function

Replaced unqualified call to log with std::log in step speed calculation to ensure correct function resolution and avoid potential compilation issues.

Update test runner options in build_and_run.sh

Changed the test runner command to use the 'console' reporter, enable verbose output, show test durations, and run tests in lexicographical order for improved test diagnostics.

Refactor Player initialization and storage handling

Added a default Player constructor for unit testing and introduced explicit init() methods for PlayerWheel and AnimusMastery to separate construction from configuration. Updated PlayerStorage to log errors for unknown reserved keys only outside of unit tests. Adjusted logger formatting for consistency. Refactored unit tests to use shared_ptr<Player> and updated test logic to match new initialization patterns.

Code format - (Clang-format)

Add unit test constructor to Player class

Introduced a dedicated Player() constructor for unit testing, initializing core components and marking the instance as a unit test mock. Added detailed documentation and usage notes to clarify its purpose and restrictions.

Code format - (Clang-format)

Set m_isUnitTestMock to true in Player constructor

Moved initialization of m_isUnitTestMock to the member initializer list in the Player default constructor for consistency and clarity.

Code format - (Clang-format)

Update CMakePresets.json

Merge branch 'dudantas/improve-storage' of https://github.com/opentibiabr/canary into dudantas/improve-storage

dudantas committed:
Remove storage access by name and related Lua bindings

Eliminates support for get/set storage values by name in C++ and Lua, including removal of Player::getStorageValueByName, PlayerStorage::get/add by name, and their Lua bindings. Updates migration scripts and talkaction to avoid using these deprecated methods, encouraging use of the KV system instead.

Remove unused upsertKey lambda from getReservedRange

Deleted the upsertKey lambda function from PlayerStorage::getReservedRange as it was not used in the function.

Refactor player storage for clarity and efficiency

Removed unnecessary 'static' from kPassThroughRanges and replaced map find with contains in has() for improved readability and efficiency.

Code format - (Clang-format)

Refactor PlayerStorage and add upsertKey method

Moved pass-through range logic into an internal namespace function for reuse and clarity. Added the upsertKey method to insert or update storage keys and mark them as modified. Improved logging and minor code cleanups in save and load methods.

Code format - (Clang-format)

dudantas committed:
Refactor player storage tests into separate functions

Each test case in player_storage_test.cpp is now encapsulated in its own static function for improved readability and maintainability. The main test suite calls these functions to register the tests.

dudantas committed:
Make onAddTileItem method const in Creature

Changed the onAddTileItem method in the Creature class to be a const member function, ensuring it does not modify the object state.

dudantas committed:
Simplify logger warning message in savePlayerStorage

Updated the logger warning in IOLoginDataSave::savePlayerStorage to remove redundant text and improve clarity when player is nullptr.

luan commented on CMakePresets.json:
```diff
@@ -58,6 +58,21 @@
+        "BUILD_TESTS": "ON"
+      }
+    },
     {
```
**note:** This is ok, but I'm changing how this works on #3677 so it'll need to be removed again.

luan commented on data-otservbr-global/scripts/game_migrations/20241708485868_move_some_storages_to_kv.lua:
```diff
@@ -8,10 +8,12 @@ local function migrate(player)
+			player:kv():set("talkaction.potions.flask", true)
+			player:setStorageValueByName(flasksStorage, -1)
+		end
 	end
```
**question:** this has been in compatibility mode with KV for a long time, do we still need `flasksStorage`? I think it could be simplified to use `kv` only.

dudantas replied:
To be honest, I don't think anyone needs it anymore, but I think it might be necessary if an "old server" wants to migrate, so I put the check to not give "nil value".

luan commented on data-otservbr-global/scripts/game_migrations/20241715984279_move_wheel_scrolls_from_storagename_to_kv.lua:
```diff
@@ -8,9 +8,11 @@ local promotionScrolls = {
+			local oldStorage = player:getStorageValueByName(scrollTable.oldScroll)
+			if oldStorage > 0 then
+				player:kv():scoped("wheel-of-destiny"):scoped("scrolls"):set(scrollTable.newScroll, true)
+			end
```
same as above, do we still need storage here?

luan commented on src/creatures/players/player.hpp:
```diff
@@ -1583,6 +1607,8 @@ class Player final : public Creature, public Cylinder, public Bankable {
 	// Bestiary
 	bool charmExpansion = false;
 
+	bool m_isUnitTestMock = false;
```
**thought:** this looks odd to me in production code, I'm not sure if this is "normal" though.

dudantas replied:
I have added this line for testing purposes. I've already removed it, thanks.

luan reviewed (approved):
Small comments if you want to address.
My main concern here is that we still have two storage systems for schemaless data. I wonder if we should come up with a way to unify this at some point. KV probably needs an easier interface if folks are still gravitating towards storage because if feels simpler. But a lot of the benefits you are gaining here KV already has (cached values, optimized & isolated db access, testability, etc)

dudantas committed:
Refactor player storage persistence to repository pattern

Replaces direct database access in PlayerStorage with a repository interface and implementation. Adds PlayerStorageRow and PlayerStorageDelta types, refactors load/save logic to ingest/delta/prepareForPersist/clearDirty methods, and updates IOLoginDataLoad/Save to use the new repository. Improves test coverage and decouples storage logic from database details.

github-actions[bot] committed:
Code format - (Clang-format)

dudantas committed:
Remove migration scripts and update player storage logic

Deleted obsolete Lua migration scripts for storage key migration. Refactored player storage to remove m_isUnitTestMock checks and updated related unit tests to use dependency injection for logging.

dudantas commented:
> Small comments if you want to address. My main concern here is that we still have two storage systems for schemaless data. I wonder if we should come up with a way to unify this at some point. KV probably needs an easier interface if folks are still gravitating towards storage because if feels simpler. But a lot of the benefits you are gaining here KV already has (cached values, optimized & isolated db access, testability, etc)

Well, the biggest problem with this is that people have been using storage for a decade or so. It's tough to "remove" the system; there are thousands of scripts in the community that use this. Of course, it would be possible to create some abstraction in the future, but I'm not sure if it's necessary at this time.

My refactoring is mainly because, with approximately 200 active players in one year, it's possible to save 5 million storage easily. A player can have thousands or hundreds of thousands of them. So, this pull request only saves what is actually modified, which, honestly, in a "high-load" scenario, is even more performant than kv as it is today. However, we could consider creating an abstraction that maintains storage usage and, under the hood, utilizes KV, but we need to think carefully about this approach. It would be a good topic for another thread.

dudantas committed:
Delete windows-test configuration preset

dudantas committed:
Refactor test env and add player storage integration tests

Extracted common test environment setup into test_env.hpp and updated account_repository_db_it.cpp to use it. Added new integration tests for DbPlayerStorageRepository with supporting CMakeLists.txt. This improves test code reuse and adds coverage for player storage repository database operations.

dudantas committed:
Refactor test utilities and player storage tests

Extracted row search logic in player storage repository integration tests into a helper function for clarity. Refactored TestEnv to separate initialization logic, improve encapsulation, and provide a global instance for easier access.

dudantas committed:
Refactor hasRow to use std::ranges::any_of

Simplified the hasRow function by replacing the manual loop with std::ranges::any_of for improved readability and modern C++ usage.

dudantas committed:
Refactor Player class member initialization order

Reordered m_playerVIP initialization in Player constructors for consistency. Removed unused m_isUnitTestMock member from Player. Renamed variable 'd' to 'delta' in savePlayerStorage for clarity.
